### ISSUE TITLE
Feature request:  Mark functions as requiring type annotations

### ISSUE THREAD

**Issue by newhoggy** [ndmitchell/hlint#616]
The function `show` not a safe function to use because it has instances on nearly every time and during refactoring we may accidentally run `show` on the wrong type by accident.

It would be very useful to have a rule that tells `hlint` to warn on functions like `show` to assert that they should always be called with type applications to make it obvious if `show` is accidentally called with an argument of the wrong type.

**Comment by AlexeyRaga** [comment]
Isn't 
```
{-# HLINT error myFunction #-}
```
what you are looking it?

**Comment by newhoggy** [comment]
No because then I can't call `myFunction` at all.

**Comment by newhoggy** [comment]
I want to make this okay:

```haskell
show @String x
```

But this fail:

```haskell
show x
```

**Comment by AlexeyRaga** [comment]
OK, what about
```
{-# HLINT warn myFunction #-}
```

**Comment by newhoggy** [comment]
Same problem.  It will warn on all calls to `myFunction`.

**Comment by ndmitchell** [comment]
I think this would be useful. At a previous `$WORK` we banned `show` without type applications for exactly this reason. You can't _quite_ do it it HLint, but you can come pretty close:

```
- warn: {lhs: read x, rhs: not allowed, side: not (isTypeApp x)}
```

Now it will detect `read x` where `x` isn't a type application and warn. You can still write `let r = read in r x`, but not the simple versions.
